diff --git a/examples/galleryapp/AndroidManifest.xml b/examples/galleryapp/AndroidManifest.xml
index 3b0ac2e..d654634 100644
--- a/examples/galleryapp/AndroidManifest.xml
+++ b/examples/galleryapp/AndroidManifest.xml
@@ -12,7 +12,7 @@
                 <action android:name="android.intent.action.MAIN"/>
                 <category android:name="android.intent.category.LAUNCHER"/>
             </intent-filter>
-            <meta-data android:name="android.app.lib_name" android:value="-- %%INSERT_APP_LIB_NAME%% --"/>
+            <meta-data android:name="android.app.lib_name" android:value="kirigami2gallery"/>
             <meta-data android:name="android.app.qt_sources_resource_id" android:resource="@array/qt_sources"/>
             <meta-data android:name="android.app.repository" android:value="default"/>
             <meta-data android:name="android.app.qt_libs_resource_id" android:resource="@array/qt_libs"/>
@@ -53,10 +53,6 @@
     <uses-sdk android:minSdkVersion="16" android:targetSdkVersion="16"/>
     <supports-screens android:largeScreens="true" android:normalScreens="true" android:anyDensity="true" android:smallScreens="true"/>
 
-    <!-- The following comment will be replaced upon deployment with default permissions based on the dependencies of the application.
-         Remove the comment if you do not require these default permissions. -->
-    <!-- %%INSERT_PERMISSIONS -->
-
     <!-- The following comment will be replaced upon deployment with default features based on the dependencies of the application.
          Remove the comment if you do not require these default features. -->
     <!-- %%INSERT_FEATURES -->
diff --git a/examples/galleryapp/CMakeLists.txt b/examples/galleryapp/CMakeLists.txt
index c3c1734..1326d27 100644
--- a/examples/galleryapp/CMakeLists.txt
+++ b/examples/galleryapp/CMakeLists.txt
@@ -37,30 +37,6 @@ else()
     message(WARNING "KPackage components should be specified in reverse domain notation. Appstream information won't be generated for ${component}.")
 endif()
 
-find_program(kpackagetool_cmd kpackagetool5)
-if (kpackagetool_cmd)
-    set(component org.kde.kirigami2.gallery)
-    set(APPDATAFILE "${CMAKE_CURRENT_BINARY_DIR}/${component}.appdata.xml")
-
-    message(STATUS "${kpackagetool_cmd} --appstream-metainfo ${CMAKE_CURRENT_SOURCE_DIR}/../gallerydata --appstream-metainfo-output ${APPDATAFILE}")
-    execute_process(
-        COMMAND ${kpackagetool_cmd} --appstream-metainfo ${CMAKE_CURRENT_SOURCE_DIR}/../gallerydata --appstream-metainfo-output ${APPDATAFILE}
-        ERROR_VARIABLE appstreamerror
-        RESULT_VARIABLE result)
-    if (NOT result EQUAL 0)
-        message(WARNING "couldn't generate metainfo for ${component}: ${appstreamerror}")
-    else()
-        if(appstreamerror)
-            message(WARNING "warnings during generation of metainfo for ${component}: ${appstreamerror}")
-        endif()
-
-        # OPTIONAL because desktop files can be NoDisplay so they render no XML.
-        install(FILES ${APPDATAFILE} DESTINATION ${KDE_INSTALL_METAINFODIR} OPTIONAL)
-    endif()
-else()
-    message(WARNING "KPackage components should be specified in reverse domain notation. Appstream information won't be generated for ${component}.")
-endif()
-
 add_executable(kirigami2gallery ${kirigami2gallery_SRCS} ${RESOURCES})
 target_link_libraries(kirigami2gallery Qt5::Core  Qt5::Qml Qt5::Quick Qt5::Svg ${kirigami2gallery_EXTRA_LIBS})
 
diff --git a/examples/gallerydata/contents/ui/DesktopExampleApp.qml b/examples/gallerydata/contents/ui/DesktopExampleApp.qml
index 56aedce..61f65b3 100644
--- a/examples/gallerydata/contents/ui/DesktopExampleApp.qml
+++ b/examples/gallerydata/contents/ui/DesktopExampleApp.qml
@@ -86,7 +86,10 @@ Kirigami.ApplicationWindow {
             },
             Kirigami.Action {
                 text: "Open A Layer"
-                iconName: "configure"
+                icon {
+                    name: "configure"
+                    color: Kirigami.Theme.activeTextColor
+                }
                 onTriggered: {
                     pageStack.layers.push(Qt.resolvedUrl("gallery/LayersGallery.qml"));
                 }
diff --git a/examples/gallerydata/contents/ui/gallery/MiscGallery.qml b/examples/gallerydata/contents/ui/gallery/MiscGallery.qml
index af4ffd9..f3a34c4 100644
--- a/examples/gallerydata/contents/ui/gallery/MiscGallery.qml
+++ b/examples/gallerydata/contents/ui/gallery/MiscGallery.qml
@@ -28,18 +28,27 @@ ScrollablePage {
 
     actions {
         main: Action {
-            iconName: "document-edit"
+            icon {
+                name: "document-edit"
+                color: Theme.negativeTextColor
+            }
             text: "Main Action Text"
         }
         left: Action {
-            iconName: "go-previous"
+            icon {
+                name: "go-previous"
+                color: Theme.positiveTextColor
+            }
             text: "Left Action Text"
             onTriggered: {
                 showPassiveNotification("Left action triggered")
             }
         }
         right: Action {
-            iconName: "go-next"
+            icon {
+                name: "go-next"
+                color: Theme.neutralTextColor
+            }
             text: "Right Action Text"
             onTriggered: {
                 showPassiveNotification("Right action triggered")
@@ -48,12 +57,15 @@ ScrollablePage {
         contextualActions: [
             Action {
                 text:"Action for buttons"
-                iconName: "bookmarks"
+                icon {
+                    name: "bookmarks"
+                    color: Theme.activeTextColor
+                }
                 onTriggered: showPassiveNotification("Action 1 clicked")
             },
             Action {
                 text:"Disabled Action"
-                iconName: "folder"
+                icon.name: "folder"
                 enabled: false
             },
             Action {
diff --git a/examples/gallerydata/contents/ui/gallery/NonScrollableGallery.qml b/examples/gallerydata/contents/ui/gallery/NonScrollableGallery.qml
index 12a40f7..9e608f6 100644
--- a/examples/gallerydata/contents/ui/gallery/NonScrollableGallery.qml
+++ b/examples/gallerydata/contents/ui/gallery/NonScrollableGallery.qml
@@ -29,7 +29,9 @@ Page {
 
     actions {
         main: Action {
-            iconName: sheet.sheetOpen ? "dialog-cancel" : "document-edit"
+            icon {
+                name: sheet.sheetOpen ? "dialog-cancel" : "document-edit"
+            }
             text: "Main Action Text"
             checkable: true
             onCheckedChanged: sheet.sheetOpen = checked;
diff --git a/examples/gallerydata/contents/ui/gallery/SliderGallery.qml b/examples/gallerydata/contents/ui/gallery/SliderGallery.qml
index 24b304b..119a768 100644
--- a/examples/gallerydata/contents/ui/gallery/SliderGallery.qml
+++ b/examples/gallerydata/contents/ui/gallery/SliderGallery.qml
@@ -55,7 +55,7 @@ ScrollablePage {
                 anchors.centerIn: parent
                 spacing: Units.smallSpacing
 
-                Label {
+                Controls.Label {
                     text: "Normal:"
                 }
                 Controls.Slider {
@@ -69,7 +69,7 @@ ScrollablePage {
                         text: slider.position.toFixed(1)
                     }
                 }
-                Label {
+                Controls.Label {
                     text: "Disabled:"
                 }
                 Controls.Slider {
@@ -78,7 +78,7 @@ ScrollablePage {
                     value: 2
                     to: 5.0
                 }
-                Label {
+                Controls.Label {
                     text: "Thickmarks:"
                 }
                 Controls.Slider {
@@ -94,7 +94,7 @@ ScrollablePage {
                     }
                 }
                 Controls.RangeSlider {}
-                Label {
+                Controls.Label {
                     text: "Vertical:"
                 }
                 RowLayout {
diff --git a/examples/gallerydata/metadata.desktop b/examples/gallerydata/metadata.desktop
index b725e9e..7b19e05 100644
--- a/examples/gallerydata/metadata.desktop
+++ b/examples/gallerydata/metadata.desktop
@@ -43,6 +43,7 @@ Comment[ca@valencia]=Mostra exemples de components del Kirigami i permet jugar-h
 Comment[da]=Viser eksempler på Kirigami-komponenter og lader dig lege med dem
 Comment[es]=Muestra ejemplos de componentes de Kirigami y le permite jugar con ellos
 Comment[eu]=Kirigami osagaien adibideak erakusten ditu eta haiekin jolasten uzten dizu
+Comment[fr]=Affiche des exemples des composants Kirigami et permet de jouer avec eux
 Comment[it]=Mostra esempi di componenti Kirigami e ti permette di giocare con essi
 Comment[nl]=Toont voorbeelden van Kirigami componenten en laat u met ze spelen
 Comment[pl]=Pokazuje przykłady składników Kirigami i umożliwia zabawę nimi
@@ -56,7 +57,8 @@ Comment[sv]=Visar exempel på Kirigami-komponenter och låter dig leka med dem
 Comment[tr]=Kirigami bileşenlerinin örneklerini gösterir ve onlarla oynamanıza izin verir
 Comment[uk]=Показує приклади компонентів Kirigami і надає вам змогу попрацювати з ними
 Comment[x-test]=xxShows examples of Kirigami components and allows you to play with themxx
-Comment[zh_CN]=显示 Kirigami 组件并允许您试玩它们
+Comment[zh_CN]=显示 Kirigami 组件实例并允许您试用它们
+Comment[zh_TW]=顯示 Kirigami 組件的範例，讓您可以玩玩看
 
 Encoding=UTF-8
 Type=Application
diff --git a/examples/simpleexamples/MultipleColumnsGallery.qml b/examples/simpleexamples/MultipleColumnsGallery.qml
index 483fb18..376336e 100644
--- a/examples/simpleexamples/MultipleColumnsGallery.qml
+++ b/examples/simpleexamples/MultipleColumnsGallery.qml
@@ -48,7 +48,7 @@ ScrollablePage {
         width: page.width
         spacing: Units.smallSpacing
 
-        Label {
+        Controls.Label {
             Layout.fillWidth: true
             wrapMode: Text.WordWrap
             text: "This page is used to test multiple columns: you can push and pop an arbitrary number of pages, each new page will have a random implicit width between 8 and 35 grid units.\nIf you enlarge the window enough, you can test how the application behaves with multiple columns."
@@ -57,7 +57,7 @@ ScrollablePage {
             Layout.minimumWidth: Units.gridUnit *2
             Layout.minimumHeight: Layout.minimumWidth
         }
-        Label {
+        Controls.Label {
             anchors.horizontalCenter: parent.horizontalCenter
             text: "Page implicitWidth: " + page.implicitWidth
         }
diff --git a/src/controls/Action.qml b/src/controls/Action.qml
index 71c0b58..96abb6b 100644
--- a/src/controls/Action.qml
+++ b/src/controls/Action.qml
@@ -18,6 +18,7 @@
  */
 
 import QtQuick 2.7
+import "private"
 
 /**
  * An item that represents an abstract Action
@@ -71,13 +72,41 @@ QtObject {
      * iconName: string
      * Sets the icon name for the action. This will pick the icon with the given name from the current theme.
      */
-    property string iconName
+    property alias iconName: iconGroup.name
 
     /**
      * iconSource: string
      * Sets the icon file or resource url for the action. Defaults to the empty URL. Use this if you want a specific file rather than an icon from the theme
      */
-    property string iconSource
+    property alias iconSource: iconGroup.source
+
+    /**
+     * metadata for the icon, such as width/height.name and source
+     * * name	This property holds the name of the icon to use.
+     *       The icon will be loaded from the platform theme.
+     *       If the icon is found in the theme, it will always be used;
+     *       even if icon.source is also set. If the icon is not found,
+     *       icon.source will be used instead.
+     *       For more information on theme icons, see QIcon::fromTheme().
+     *
+     * * source	This property holds the name of the icon to use.
+     *       The icon will be loaded as a regular image.
+     *       If icon.name is set and refers to a valid theme icon,
+     *       it will always be used instead of this property.
+     *
+     * * width	This property holds the width of the icon.
+     *   The icon's width will never exceed this value,
+     *   though it will shrink when necessary.
+     *   height	This property holds the height of the icon.
+     *   The icon's height will never exceed this value,
+     *   though it will shrink when necessary.
+     *
+     * *color	This property holds the color of the icon.
+     *   The icon is tinted with the specified color, unless the color is set to "transparent".
+     */
+    property ActionIconGroup icon: ActionIconGroup {
+        id: iconGroup
+    }
 
     /**
      * shortcut : keysequence
diff --git a/src/controls/BasicListItem.qml b/src/controls/BasicListItem.qml
index b652549..3ea9f75 100644
--- a/src/controls/BasicListItem.qml
+++ b/src/controls/BasicListItem.qml
@@ -40,11 +40,12 @@ AbstractListItem {
 
     /**
      * icon: var
-     * A single icon that will be displayed in the list item. The icon can
-     * be either a QIcon, a string name of a fdo compatible name,
+     * A single icon that will be displayed in the list item.
+     * The icon can be a grouped property with name,size,color etc, as QtQuickControls2 icons are defined.
+     * The icon can also be either a QIcon, a string name of a fdo compatible name,
      * or any url accepted by the Image element.
      */
-    property alias icon: iconItem.source
+    property var icon
 
     /**
      * reserveSpaceForIcon: bool
@@ -62,10 +63,12 @@ AbstractListItem {
         property bool indicateActiveFocus: listItem.pressed || Settings.isMobile || listItem.activeFocus || (listItem.ListView.view ? listItem.ListView.view.activeFocus : false)
         Icon {
             id: iconItem
+            source: listItem.icon && listItem.icon.hasOwnProperty && listItem.icon.hasOwnProperty("name") ? listItem.icon.name : listItem.icon
             Layout.minimumHeight: Units.iconSizes.smallMedium
             Layout.maximumHeight: Layout.minimumHeight
             Layout.minimumWidth: height
             selected: layout.indicateActiveFocus && (listItem.checked || listItem.pressed)
+            color: listItem.icon && listItem.icon.color && listItem.icon.color.a > 0 ? listItem.icon.color : Qt.rgba(0, 0, 0, 0)  
         }
         QQC2.Label {
             id: labelItem
diff --git a/src/controls/ContextDrawer.qml b/src/controls/ContextDrawer.qml
index d101c0a..740b72f 100644
--- a/src/controls/ContextDrawer.qml
+++ b/src/controls/ContextDrawer.qml
@@ -133,7 +133,7 @@ OverlayDrawer {
             }
             delegate: BasicListItem {
                 checked: modelData.checked
-                icon: modelData.iconName
+                icon: modelData.icon
                 supportsMouseEvents: true
                 separatorVisible: false
                 label: model ? (model.tooltip ? model.tooltip : model.text) : (modelData.tooltip ? modelData.tooltip : modelData.text)
diff --git a/src/controls/GlobalDrawer.qml b/src/controls/GlobalDrawer.qml
index bd18401..11f3109 100644
--- a/src/controls/GlobalDrawer.qml
+++ b/src/controls/GlobalDrawer.qml
@@ -405,7 +405,7 @@ OverlayDrawer {
                                 id: listItem
                                 supportsMouseEvents: true
                                 checked: modelData.checked
-                                icon: modelData.iconName
+                                icon: modelData.icon
                                 label: modelData.text
                                 separatorVisible: false
                                 visible: model ? model.visible || model.visible===undefined : modelData.visible
diff --git a/src/controls/ScrollablePage.qml b/src/controls/ScrollablePage.qml
index 464e1b6..9f03933 100644
--- a/src/controls/ScrollablePage.qml
+++ b/src/controls/ScrollablePage.qml
@@ -108,10 +108,10 @@ Page {
         z: 0
         //child of root as it shouldn't have margins
         parent: root
-        topPadding: (applicationWindow() && applicationWindow().header ? applicationWindow().header.preferredHeight : 0) + (contentItem == flickable ? 0 : root.topPadding)
+        topPadding: (applicationWindow() && applicationWindow().header ? applicationWindow().header.preferredHeight : 0) + (contentItem == flickableItem ? 0 : root.topPadding)
         leftPadding: root.leftPadding
         rightPadding: root.rightPadding
-        bottomPadding: contentItem == flickable ? 0 : root.bottomPadding
+        bottomPadding: contentItem == flickableItem ? 0 : root.bottomPadding
         anchors {
             fill: parent
             topMargin: root.header ? root.header.height : 0
diff --git a/src/controls/ToolBarApplicationHeader.qml b/src/controls/ToolBarApplicationHeader.qml
index 76926ab..9b7033f 100644
--- a/src/controls/ToolBarApplicationHeader.qml
+++ b/src/controls/ToolBarApplicationHeader.qml
@@ -153,7 +153,7 @@ ApplicationHeader {
                     model: page && page.actions.contextualActions ? page.actions.contextualActions : null
                     delegate: BasicListItem {
                         text: modelData ? modelData.text : ""
-                        icon: modelData.iconName
+                        icon: modelData.icon
                         checkable:  modelData.checkable
                         checked: modelData.checked
                         onClicked: {
diff --git a/src/controls/private/ActionButton.qml b/src/controls/private/ActionButton.qml
index 7cffe9c..d8bf03d 100644
--- a/src/controls/private/ActionButton.qml
+++ b/src/controls/private/ActionButton.qml
@@ -231,7 +231,8 @@ Item {
                     width: height
                     visible: root.action
                     readonly property bool pressed: root.action && ((mouseArea.buttonPressedUnderMouse && mouseArea.pressed) || root.action.checked)
-                    color: pressed ? Qt.darker(Theme.highlightColor, 1.3) : Theme.highlightColor
+                    property color baseColor: root.action && root.action.icon && root.action.icon.color && root.action.icon.color != undefined && root.action.icon.color.a > 0 ? root.action.icon.color : Theme.highlightColor
+                    color: pressed ? Qt.darker(baseColor, 1.3) : baseColor
 
                     Icon {
                         id: icon
@@ -270,7 +271,8 @@ Item {
                     visible: root.leftAction
 
                     readonly property bool pressed: root.leftAction && ((mouseArea.leftButtonPressedUnderMouse && mouseArea.pressed) || root.leftAction.checked)
-                    color: pressed ? Theme.highlightColor : Theme.backgroundColor
+                    property color baseColor: root.leftAction && root.leftAction.icon && root.leftAction.icon.color && root.leftAction.icon.color != undefined && root.leftAction.icon.color.a > 0 ? root.leftAction.icon.color : Theme.highlightColor
+                    color: pressed ? baseColor : Theme.backgroundColor
                     Behavior on color {
                         ColorAnimation {
                             duration: Units.longDuration
@@ -304,7 +306,8 @@ Item {
                     width: height + (root.action ? Units.gridUnit*2 : 0)
                     visible: root.rightAction
                     readonly property bool pressed: root.rightAction && ((mouseArea.rightButtonPressedUnderMouse && mouseArea.pressed) || root.rightAction.checked)
-                    color: pressed ? Theme.highlightColor : Theme.backgroundColor
+                    property color baseColor: root.rightAction && root.rightAction.icon && root.rightAction.icon.color && root.rightAction.icon.color != undefined && root.rightAction.icon.color.a > 0 ? root.rightAction.icon.color : Theme.highlightColor
+                    color: pressed ? baseColor : Theme.backgroundColor
                     Behavior on color {
                         ColorAnimation {
                             duration: Units.longDuration
diff --git a/src/controls/private/ActionIconGroup.qml b/src/controls/private/ActionIconGroup.qml
new file mode 100644
index 0000000..19a5ea1
--- /dev/null
+++ b/src/controls/private/ActionIconGroup.qml
@@ -0,0 +1,29 @@
+/*
+ *   Copyright 2017 Marco Martin <mart@kde.org>
+ *
+ *   This program is free software; you can redistribute it and/or modify
+ *   it under the terms of the GNU Library General Public License as
+ *   published by the Free Software Foundation; either version 2, or
+ *   (at your option) any later version.
+ *
+ *   This program is distributed in the hope that it will be useful,
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *   GNU Library General Public License for more details
+ *
+ *   You should have received a copy of the GNU Library General Public
+ *   License along with this program; if not, write to the
+ *   Free Software Foundation, Inc.,
+ *   51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+import QtQml 2.1
+
+QtObject {
+    property string name
+    property string source
+    property int width
+    property int height
+    property color color: Qt.rgba(0, 0, 0, 0)
+}
+
diff --git a/src/controls/private/PrivateActionToolButton.qml b/src/controls/private/PrivateActionToolButton.qml
index 4488931..4a9c68d 100644
--- a/src/controls/private/PrivateActionToolButton.qml
+++ b/src/controls/private/PrivateActionToolButton.qml
@@ -57,10 +57,11 @@ Controls.ToolButton {
             Icon {
                 Layout.minimumWidth: 22
                 Layout.minimumHeight: 22
-                source: control.action ? control.action.iconName : ""
+                source: control.action ? (control.action.icon ? control.action.icon.name : control.action.iconName) : ""
                 visible: control.action && control.action.iconName != ""
+                color: control.action && control.action.icon && control.action.icon.color.a > 0 ? control.action.icon.color : Qt.rgba(0, 0, 0, 0)                                                                                     
             }
-            Label {
+            Controls.Label {
                 text: action ? action.text : ""
                 visible: control.showText
             }
diff --git a/src/libkirigami/platformtheme.cpp b/src/libkirigami/platformtheme.cpp
index 4846af3..a8d3995 100644
--- a/src/libkirigami/platformtheme.cpp
+++ b/src/libkirigami/platformtheme.cpp
@@ -67,6 +67,7 @@ public:
     QFont font;
     QPalette palette;
     bool m_inherit = true;
+    bool m_init = true;
 };
 
 PlatformThemePrivate::PlatformThemePrivate(PlatformTheme *q)
@@ -129,6 +130,7 @@ PlatformTheme::PlatformTheme(QObject *parent)
             d->findParentStyle();
         });
     }
+    d->m_init = false;
     //TODO: needs https://codereview.qt-project.org/#/c/206889/ for font changes
 }
 
@@ -153,8 +155,10 @@ void PlatformTheme::setColorSet(PlatformTheme::ColorSet colorSet)
         }
     }
 
-    emit colorSetChanged(colorSet);
-    d->setColorCompressTimer->start();
+    if (!d->m_init) {
+        emit colorSetChanged(colorSet);
+        d->setColorCompressTimer->start();
+    }
 }
 
 PlatformTheme::ColorSet PlatformTheme::colorSet() const
diff --git a/src/libkirigami/platformtheme.h b/src/libkirigami/platformtheme.h
index cb5fc70..044c18d 100644
--- a/src/libkirigami/platformtheme.h
+++ b/src/libkirigami/platformtheme.h
@@ -148,6 +148,7 @@ public:
         Tooltip, /** Color set used by tooltips */
         Complementary /** Color set meant to be complementary to Window: usually is a dark theme for light themes */
     };
+    Q_ENUM(ColorSet)
 
     explicit PlatformTheme(QObject *parent = 0);
     ~PlatformTheme();
